### PR TITLE
Refactor/pde spec

### PR DESCRIPTION
**Pull request by stonecoldhughes**
Help! I cannot get this to compile.

### CODE REVIEW

**Comment by stonecoldhughes** [comment]
> Help! I cannot get this to compile.

File: src/pde/pde_base.hpp [hunk 1]
```diff
@@ -428,12 +444,14 @@ class PDE
...
+  term_set<P> &get_terms() { return terms_; }
 
   fk::matrix<P> const &get_coefficients(int const term, int const dim) const
   {
     return terms_[term][dim].get_coefficients();
   }
+
+  /* Is this function used anywhere? It doesn't really look like it */
   void
   set_coefficients(fk::matrix<P> const coeffs, int const term, int const dim)
```
Comment (dlg0) [inline src/pde/pde_base.hpp]: does this exist in the matlab? how did it end up here in the first place?

File: src/pde/pde_continuity1.hpp [hunk 2]
```diff
@@ -112,30 +112,30 @@ class PDE_continuity_1d : public PDE<P>
...
+      dimension<P>(-1.0,                   // domain min
+                   1.0,                    // domain max
+                   2,                      // levels
+                   2,                      // degree
+                   initial_condition_dim0, // initial condition
+                   "x");                   // name
 
   inline static std::vector<dimension<P>> const dimensions_ = {dim0_};
 
+  inline static class partial_term<P> partial_term_0 = partial_term<P>(
```
Comment (dlg0) [inline src/pde/pde_continuity1.hpp]: Can we format this vertically aligned and add clang-format ignore?

File: src/coefficients.cpp [hunk 3]
```diff
@@ -12,50 +12,31 @@
...
-    // Equation 2 of LDG
+  /* unroll the loop one iteration */
+  fk::matrix<double> full_operator =
+      partial_operator(dim, term_1D, time, rotate, term_1D.get_partial_term(0));
 
-    dimension<P> const dim_2(term_1D.BC_left_2, term_1D.BC_right_2,
-                             dim.domain_min, dim.domain_max, dim.get_level(),
-                             dim.get_degree(), dim.initial_condition, dim.name);
+  /* invoke the type conversion constructor */
+  term_1D.get_partial_term(0).set_coefficients(fk::matrix<P>(full_operator));
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: reason for the `(0)`: there are `num_pterms` partial terms that must be chained together to form the full operator. To start the chain, we must first create the first link. The full_operator in this first step of construction only consists of one link - the first partial term. It does not need to be multiplied by the product of all previous partial terms because there are none yet. After the first link is established, the following loop starts at index 1. It's like a loop that started at index 0 that has been unrolled 1 iteration and the first link in the chain is handled differently than all that will follow

[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: reason why the coefficient matrix for the partial term is not done inside of `partial_operator()`:
if the coefficient matrix for the partial term is not set within partial_operator(), the partial_term can be a const reference. Setting the partial term's coefficient matrix is a very small but very important step, and if it happened as a quick side adventure inside of a function as extensive and complex as partial_operator(), it might effectively be hidden and discovering the origin of the partial terms could be difficult for future developers. `partial_operator()` has one job - return the coefficient_matrix. However, if you believe that it should be done inside that function, I will do that.

[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: Should `partial_operator` be a method of the `term_1D` class:
I think this question boils down to: "Should this be programmed with a functional paradigm or an object-oriented one?" If you had asked me yesterday, I would have agreed that it should be in the class. However, I rewrite my mpi module in a functional paradigm and reduced the code size significantly while also making it clearer. Furthermore, the rest of the codebase seems to be functional rather than object-oriented. I cite Tyler as my source. To me it seems clearer this way, but if that's simply not true I can move it into the class.

[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: difference between `partial_operator()` and `get_coefficients()`:
`partial_operator()` creates the coefficients matrix based on the partial term's generator function g_func as well as the boundary conditions and coefficient types.
`get_coefficients()` simply returns the data that had been previously generated by a call to `partial_operator()`. Perhaps a different function name such as `generate_partial_operator_matrix()` would be a better name.

File: src/pde/pde_base.hpp [hunk 4]
```diff
@@ -149,6 +149,70 @@ enum class flux_type
...
+    {
+      flux_scale = 0.0;
+    }
+
+    return;
+  }
+
+  void set_flux_scale(P const dfdu)
+  {
+    assert(flux == flux_type::lax_friedrich);
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: Since I didn't leave myself a comment, I have no idea why I did that. No, none of the existing PDE's use the lax_friedrich type. I will take it out.

File: src/pde/pde_base.hpp [hunk 5]
```diff
@@ -428,12 +444,14 @@ class PDE
   {
     return dimensions_;
   }
-  term_set<P> const &get_terms() const { return terms_; }
+  term_set<P> &get_terms() { return terms_; }
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: the reference is to avoid making a copy of the whole term_set. It was changed from const to non-const because the term_set contains terms, which contains partial_terms, which have to be modified because their partial coefficient operator matrices must be set.

File: src/pde/pde_base.hpp (hunk 1 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: my comment is stupid - turns out that the function is used in main.cpp as well as in a slew of test files. I took it out

File: src/pde/pde_continuity1.hpp (hunk 2 above)
Comment (stonecoldhughes) [reply src/pde/pde_continuity1.hpp]: absolutely. Would you like this formatting to likewise be applied across all the test PDEs?

File: src/coefficients.cpp [hunk 6]
```diff
@@ -12,50 +12,31 @@
...
-    // Equation 1 of LDG
-
-    dimension<P> const dim_1(term_1D.BC_left_1, term_1D.BC_right_1,
-                             dim.domain_min, dim.domain_max, dim.get_level(),
-                             dim.get_degree(), dim.initial_condition, dim.name);
-
-    term<P> const term_1D_1(coefficient_type::grad, term_1D.g_func_1,
-                            term_1D.time_dependent, term_1D.flux_1,
-                            term_1D.get_data(), term_1D.name, dim_1);
+  int const len = term_1D.n_partial_terms();
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: done

File: src/coefficients.cpp [hunk 7]
```diff
@@ -12,50 +12,31 @@
...
-                             dim.get_degree(), dim.initial_condition, dim.name);
+  /* invoke the type conversion constructor */
+  term_1D.get_partial_term(0).set_coefficients(fk::matrix<P>(full_operator));
 
-    term<P> const term_1D_2(coefficient_type::grad, term_1D.g_func_2,
-                            term_1D.time_dependent, term_1D.flux_2,
-                            term_1D.get_data(), term_1D.name, dim_2);
+  for (int i = 1; i < len; i++)
+  {
+    fk::matrix<double> accumulator = partial_operator(
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: done

File: src/pde/pde_base.hpp [hunk 8]
```diff
@@ -60,20 +60,20 @@ template<typename P>
 class dimension
 {
 public:
+  /*
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: done

File: src/pde/pde_base.hpp [hunk 9]
```diff
@@ -248,26 +270,21 @@ class term
     return d.get_degree() * static_cast<int>(std::pow(2, d.get_level()));
   };
 
+  class partial_term<P> &get_partial_term(int i)
+  {
+    return partial_terms[i];
+  };
+
+  int n_partial_terms() const { return partial_terms.size(); };
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: done

File: src/coefficients.cpp [hunk 10]
```diff
@@ -64,9 +46,9 @@ generate_coefficients(dimension<P> const &dim, term<P> const &term_1D,
 // coefficient matricies
 template<typename P>
 fk::matrix<double>
-generate_mass_or_grad_coefficients(dimension<P> const &dim,
-                                   term<P> const &term_1D, double const time,
-                                   bool const rotate)
+generate_partial_operator_matrix(dimension<P> const &dim, term<P> &term_1D,
+                                 double const time, bool const rotate,
+                                 class partial_term<P> const &partial_term)
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: done

File: src/pde.hpp [hunk 11]
```diff
@@ -89,19 +89,6 @@ make_PDE(PDE_opts choice, int const level = -1, int const degree = -1)
     return std::make_unique<PDE_fokkerplanck_1d_4p1a<P>>(level, degree);
   case PDE_opts::fokkerplanck_1d_4p2:
     return std::make_unique<PDE_fokkerplanck_1d_4p2<P>>(level, degree);
-  // TODO not yet implemented, replace return with appropriate types
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde.hpp]: this was a copy paste error

File: src/pde/pde_base.hpp [hunk 12]
```diff
@@ -361,6 +369,12 @@ class PDE
       assert(exact_vector_funcs.size() == static_cast<unsigned>(num_dims));
     }
 
+    // check all terms
+    for (std::vector<term<P>> const term_list : terms_)
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: good catch

File: src/pde/pde_base.hpp [hunk 13]
```diff
@@ -149,6 +145,66 @@ enum class flux_type
...
+class partial_term
+{
+public:
+  partial_term(coefficient_type const coeff_type, g_func_type<P> const g_func,
+               flux_type const flux, boundary_condition const left,
+               boundary_condition const right)
+      : coeff_type(coeff_type), g_func(g_func), flux(flux), left(left),
+        right(right)
+  {
+    if (flux == flux_type::central)
```
[the remark this answers is not in the document]
Comment (dlg0) [reply src/pde/pde_base.hpp]: There should only be a few of these.

**Comment by dlg0** [comment]
@stonecoldhughes is this ready for @bmcdanie to review again?

File: src/coefficients.cpp [hunk 14]
```diff
@@ -12,50 +12,32 @@
 // operators
 template<typename P>
 fk::matrix<double>
-generate_coefficients(dimension<P> const &dim, term<P> const &term_1D,
+generate_coefficients(dimension<P> const &dim, term<P> &term_1D,
```
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/coefficients.cpp]: why?

File: src/coefficients.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/coefficients.cpp]: this discussion is kind of hard to follow, maybe not least because `partial_operator()` is a name that doesn't tell me much of anything. Also, if it is too big, then we should refactor.

In general, I agree that simpler and clearer is the ultimate goal; I have also observed that functional themes can aid with this.

However, I also like RAII, and so if we have acquired a "partial_term", why is it yet uninitialized?

File: src/coefficients.cpp [hunk 15]
```diff
@@ -64,9 +46,9 @@ generate_coefficients(dimension<P> const &dim, term<P> const &term_1D,
 // coefficient matricies
 template<typename P>
 fk::matrix<double>
-generate_mass_or_grad_coefficients(dimension<P> const &dim,
-                                   term<P> const &term_1D, double const time,
-                                   bool const rotate)
+generate_partial_operator_matrix(dimension<P> const &dim, term<P> &term_1D,
+                                 double const time, bool const rotate,
+                                 partial_term<P> const &partial_term)
```
Comment (grahamlopez) [inline src/coefficients.cpp]: I generally dislike variables having the same name as their type. Anyone else?

File: src/pde/pde_base.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/pde/pde_base.hpp]: instead of an enum class, why not just let the `flux_type` hold these values and set it directly?

File: src/coefficients.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: This thread is a little outdated at this point, I changed the name to `generate_partial_operator_matrix`. The partial term follows the same pattern that the term did originally. It is initialized at construction with the information needed to generate the coefficients, but they are not explicitly generated on construction.

File: src/pde/pde_base.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: Terms were initialized with `g_function` and other stuff to be used for generating a coefficient matrix later. I'm not sure why it happens later and not at construction. That was a pattern already baked into main(). We would have to ask David how and if that can be changed.

[the remark this answers is not in the document]
Comment (dlg0) [reply src/pde/pde_base.hpp]: Yes, RAII ... but let’s create a separate issue for that since I want to get this PR merged and not let its scope creep.

[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: I will change it back and have it just return data. I can update the pde's setter to do what I deconsted it for originally.

File: src/pde/pde_base.hpp [hunk 16]
```diff
@@ -248,26 +262,21 @@ class term
     return d.get_degree() * static_cast<int>(std::pow(2, d.get_level()));
   };
 
+  class partial_term<P> &get_partial_term(int i)
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: done

File: src/pde/pde_continuity1.hpp [hunk 17]
```diff
@@ -112,30 +112,35 @@ class PDE_continuity_1d : public PDE<P>
...
+                   1.0,                    // domain max
+                   2,                      // levels
+                   2,                      // degree
+                   initial_condition_dim0, // initial condition
+                   "x");                   // name
 
   inline static std::vector<dimension<P>> const dimensions_ = {dim0_};
 
+  // clang-format off
+  inline static class partial_term<P>
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_continuity1.hpp]: done

File: src/pde/pde_continuity2.hpp [hunk 18]
```diff
@@ -183,67 +183,71 @@ class PDE_continuity_2d : public PDE<P>
...
-              fk::vector<P>(),        // additional data vector
-              "v_x.d_dx",             // name
-              dim0_);                 // owning dim
+      term<P>(false,           // time-dependent
+              fk::vector<P>(), // additional data vector
+              "v_x.d_dx",      // name
+              dim0_,           // owning dim
+              {partial_term_t0_d0});
+
+  inline static class partial_term<P> partial_term_t0_d1 = partial_term<P>(
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_continuity2.hpp]: done

File: src/coefficients.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: didn't consider whether it was bad design or not at the time but it's so that it can secretly be a setter function in addition to generating coefficients... I will fix that

File: src/coefficients.cpp (hunk 15 above)
Comment (stonecoldhughes) [reply src/coefficients.cpp]: I'll come up with something more creative...

File: src/coefficients.cpp [hunk 19]
```diff
@@ -12,50 +12,32 @@
...
-  {
-    auto coefficients =
-        generate_mass_or_grad_coefficients(dim, term_1D, time, rotate);
 
-    return coefficients;
-  }
+  return full_operator;
 }
 
 // construct 1D coefficient matrix - new conventions
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: I can always count on Gcc (Graham compiler collection) when gcc fails to find these errors.

File: src/pde/pde_base.hpp [hunk 20]
```diff
@@ -149,6 +145,66 @@ enum class flux_type
...
+    }
+    else
+    {
+      flux_scale = 0.0;
+    }
+
+    return;
+  }
+
+  void set_flux_scale(P const dfdu) { flux_scale = dfdu; };
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: correct, that is not used anywhere in the codebase. That would be a David question... perhaps it will be used for an upcoming feature?

File: src/pde/pde_base.hpp [hunk 21]
```diff
@@ -149,6 +145,66 @@ enum class flux_type
...
+
+  g_func_type<P> const g_func;
+
+  flux_type const flux;
+
+  boundary_condition const left;
+
+  boundary_condition const right;
+
+  P flux_scale;
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: no idea... I named the branch `refactor` but it probably should have been a feature because I didn't even look at these things. I will note them for an actual refactor later.

File: src/pde/pde_base.hpp [hunk 22]
```diff
@@ -149,6 +145,66 @@ enum class flux_type
...
+
+  P get_flux_scale() const { return flux_scale; };
+
+  coefficient_type const coeff_type;
+
+  g_func_type<P> const g_func;
+
+  flux_type const flux;
+
+  boundary_condition const left;
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: they are set in the class constructor. Not sure why it didn't show up... maybe I hadn't pushed or something

File: src/pde/pde_continuity1.hpp [hunk 23]
```diff
@@ -112,30 +112,35 @@ class PDE_continuity_1d : public PDE<P>
...
-
-  inline static term_set<P> const terms_ = {terms0_};
+  inline static term<P> term0_dim0_ =
+      term<P>(false,           // time-dependent
+              fk::vector<P>(), // additional data vector
+              "d_dx",          // name
+              dim0_,           // owning dim
+              {partial_term_0});
+
+  inline static std::vector<term<P>> terms0_ = {term0_dim0_};
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_continuity1.hpp]: all instances of deconsting happened for the same reason... so that a function that generated a matrix could also behave as a setter function for intermediate results. I will go back and fix these

File: src/pde/pde_base.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: for generating PDEs, the flux_type::upwind makes more sense than just 1.0 I think. So perhaps a mapping is a better way.

File: src/coefficients.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/coefficients.cpp]: just `generate_partial_operator` is fine

File: src/pde/pde_base.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/pde/pde_base.hpp]: yes, I agree names are good. I meant instead of
```
if (flux == flux_type::upwind)
  flux_scale = value;
...
``
we could just store actual values in `flux_type` so that we can be direct e.g.
```
flux_scale = flux;
````
where `flux` has been set to `flux_type::upwind` (or `flux_type.upwind` depending on your preference)

File: src/pde/pde_base.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (dlg0) [reply src/pde/pde_base.hpp]: I thought I responded to this already, apparently not.

The only place the `flux` values should be set is on PDE initialization I think, so we likely do not need this setter.

File: src/pde/pde_base.hpp [hunk 24]
```diff
@@ -146,8 +142,65 @@ enum class flux_type
...
+  void set_coefficients(fk::matrix<P> const &new_coefficients)
+  {
+    this->coefficients_.clear_and_resize(
+        new_coefficients.nrows(), new_coefficients.ncols()) = new_coefficients;
+
+    return;
+  }
+
+private:
+  fk::matrix<P> coefficients_;
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: #168 . In my opinion outside the scope of this PR, as terms have always had this problem, but let's fix it.

File: src/pde/pde_base.hpp [hunk 25]
```diff
@@ -161,30 +214,11 @@ class term
...
-       boundary_condition const BC_left_2  = boundary_condition::neumann,
-       boundary_condition const BC_right_2 = boundary_condition::neumann)
-      : coeff_type(coeff_type), g_func(g_func), time_dependent(time_dependent),
-        flux(flux), name(name), owning_dim(owning_dim), g_func_1(g_func_1),
-        g_func_2(g_func_2), flux_1(flux_1), flux_2(flux_2),
-        BC_left_1(BC_left_1), BC_right_1(BC_right_1), BC_left_2(BC_left_2),
-        BC_right_2(BC_right_2), data_(data)
+  term(bool const time_dependent, fk::vector<P> const data,
+       std::string const name, dimension<P> const owning_dim,
+       std::initializer_list<partial_term<P>> const partial_terms = {})
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: I don't think a term without partial terms can be constructed/is valid in our current framework.

I like use of initializer list here. remove default arg.

File: src/pde/pde_base.hpp [hunk 26]
```diff
@@ -241,6 +252,13 @@ class term
         .transfer_from(new_coefficients);
   }
 
+  void set_partial_coefficients(fk::matrix<P> const &coeffs, int const pterm)
+  {
+    assert(partial_terms.size() > 0);
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: remove this and assert that the argument falls within container bounds.

`assert(pterm >= 0)`
`assert(pterm < static_cast<int>(partial_terms.size())`

File: src/pde/pde_base.hpp [hunk 27]
```diff
@@ -366,6 +378,12 @@ class PDE
       assert(exact_vector_funcs.size() == static_cast<unsigned>(num_dims));
     }
 
+    // check all terms
+    for (std::vector<term<P>> const &term_list : terms_)
+    {
+      assert(term_list.size() == static_cast<unsigned>(num_dims));
+    }
+
     // modify for appropriate level/degree
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: loop over partial terms of each term and make sure they're nonempty

File: src/pde/pde_continuity1.hpp [hunk 28]
```diff
@@ -112,30 +112,35 @@ class PDE_continuity_1d : public PDE<P>
...
+                   1.0,                    // domain max
+                   2,                      // levels
+                   2,                      // degree
+                   initial_condition_dim0, // initial condition
+                   "x");                   // name
 
   inline static std::vector<dimension<P>> const dimensions_ = {dim0_};
 
+  // clang-format off
+  inline static partial_term<P>
```
Comment (bmcdanie) [inline src/pde/pde_continuity1.hpp]: `const` here and throughout

File: src/pde/pde_continuity1.hpp (hunk 23 above)
Comment (bmcdanie) [inline src/pde/pde_continuity1.hpp]: const here

File: src/pde/pde_base.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: #168

File: src/batch_tests.cpp [hunk 29]
```diff
@@ -2688,6 +2691,7 @@ TEMPLATE_TEST_CASE("batch builder", "[batch]", float, double)
         pde->set_coefficients(coeffs, j, i);
       }
     }
+    */
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/batch_tests.cpp]: done

File: src/batch_tests.cpp [hunk 30]
```diff
@@ -2744,6 +2748,8 @@ TEMPLATE_TEST_CASE("batch builder", "[batch]", float, double)
 
     element_table const elem_table(o, pde->num_dims);
 
+    generate_all_coefficients(pde);
+    /*
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/batch_tests.cpp]: done

File: src/batch_tests.cpp [hunk 31]
```diff
@@ -2812,6 +2819,8 @@ TEMPLATE_TEST_CASE("batch builder", "[batch]", float, double)
...
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/batch_tests.cpp]: done

File: src/coefficients.cpp [hunk 32]
```diff
@@ -8,65 +8,52 @@
 #include "transformations.hpp"
 #include <numeric>
 
-// wrap generate_coefficients to allow for construction of the diffusion type
-// operators
 template<typename P>
-fk::matrix<double>
-generate_coefficients(dimension<P> const &dim, term<P> const &term_1D,
-                      double const time, bool const rotate)
+void generate_all_coefficients(std::unique_ptr<PDE<P>> &pde, double const time,
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: done

File: src/coefficients.cpp [hunk 33]
```diff
@@ -8,65 +8,52 @@
...
-    dimension<P> const dim_2(term_1D.BC_left_2, term_1D.BC_right_2,
-                             dim.domain_min, dim.domain_max, dim.get_level(),
-                             dim.get_degree(), dim.initial_condition, dim.name);
+      /* set the partial term's coefficient matrix */
+      pde->set_partial_coefficients(j, i, 0, fk::matrix<P>(term_coeff));
 
-    term<P> const term_1D_2(coefficient_type::grad, term_1D.g_func_2,
-                            term_1D.time_dependent, term_1D.flux_2,
-                            term_1D.get_data(), term_1D.name, dim_2);
+      for (int k = 1; k < (int)partial_terms.size(); k++)
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: done

File: src/coefficients.cpp [hunk 34]
```diff
@@ -8,65 +8,52 @@
...
+      {
+        fk::matrix<double> partial_term_coeff = generate_coefficients<P>(
+            dim, term_1D, partial_terms[k], time, rotate);
 
-    auto const coefficients_2 =
-        generate_mass_or_grad_coefficients(dim_2, term_1D_2, time, rotate);
+        term_coeff = term_coeff * partial_term_coeff;
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: this won't save any copies in this case, and the * operator is more succinct

File: src/coefficients.cpp [hunk 35]
```diff
@@ -8,65 +8,52 @@
...
 
-    return coefficients_1 * coefficients_2;
-  }
-  else
-  {
-    auto coefficients =
-        generate_mass_or_grad_coefficients(dim, term_1D, time, rotate);
+        pde->set_partial_coefficients(j, i, k,
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.cpp]: Limit caffeine intake, should help

File: src/coefficients.hpp [hunk 36]
```diff
@@ -2,15 +2,36 @@
...
+                      partial_term<P> const &pterm, double const time = 0.0,
+                      bool const rotate = true);
 
 extern template fk::matrix<double>
-generate_coefficients(dimension<float> const &dim, term<float> const &term_1D,
-                      double const time = 0.0, bool const rotate = true);
+generate_coefficients<float>(dimension<float> const &dim,
+                             term<float> const &term_1D,
+                             partial_term<float> const &pterm,
+                             double const time = 0.0, bool const rotate = true);
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/coefficients.hpp]: I changed it

File: src/pde/pde_base.hpp (hunk 24 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: I'd like to save that for issue-thon

File: src/pde/pde_base.hpp (hunk 26 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: done

File: src/pde/pde_base.hpp (hunk 27 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: done

File: src/pde/pde_continuity1.hpp (hunk 28 above)
Comment (stonecoldhughes) [reply src/pde/pde_continuity1.hpp]: done

File: src/pde/pde_continuity1.hpp (hunk 23 above)
Comment (stonecoldhughes) [reply src/pde/pde_continuity1.hpp]: done

File: src/coefficients.cpp (hunk 34 above)
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/coefficients.cpp]: *without using more space. we can revisit if this becomes an issue for time-dependent matrices.

File: src/coefficients.hpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/coefficients.hpp]: great - please change time to P for now per david yesterday.

File: src/pde/pde_base.hpp (hunk 24 above)
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: yeah - that's why I opened the issue

File: src/pde/pde_base.hpp (hunk 26 above)
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: does not appear to be fixed

File: src/pde/pde_base.hpp [hunk 37]
```diff
@@ -251,28 +269,23 @@ class term
...
     return d.get_degree() * static_cast<int>(std::pow(2, d.get_level()));
-  };
+  }
+
+  std::vector<partial_term<P>> const &get_partial_terms() const
+  {
+    return partial_terms;
+  }
+
+  int num_partial_terms() const { return partial_terms.size(); };
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: is this needed since we have `get_partial_terms` above?

File: src/pde/pde_base.hpp [hunk 38]
```diff
@@ -366,6 +378,17 @@ class PDE
...
     }
 
+    // check all terms
+    for (std::vector<term<P>> const &term_list : terms_)
+    {
+      assert(term_list.size() == static_cast<unsigned>(num_dims));
+
+      for (term<P> const &term_1D : term_list)
+      {
+        assert(term_1D.get_partial_terms().size() > 0);
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: thanks.

File: src/pde/pde_base.hpp (hunk 26 above)
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: fixed

File: src/pde/pde_base.hpp (hunk 25 above)
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: fixed

File: src/coefficients.hpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/coefficients.hpp]: fixed

File: src/pde/pde_base.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/pde/pde_base.hpp]: fixed.

File: src/coefficients.cpp [hunk 39]
```diff
@@ -8,65 +8,53 @@
...
 
-// wrap generate_coefficients to allow for construction of the diffusion type
-// operators
+/* generate coefficient matrices for each 1D term in each dimension and
+   underlying partial term coefficients matrices */
 template<typename P>
-fk::matrix<double>
-generate_coefficients(dimension<P> const &dim, term<P> const &term_1D,
-                      double const time, bool const rotate)
+void generate_all_coefficients(PDE<P> &pde, P const time, bool const rotate)
```
Comment (grahamlopez) [inline src/coefficients.cpp]: why make this by reference, when a PDE<T> must be a pointer in our current design anyway?

Comment (grahamlopez) [inline src/coefficients.cpp]: why make this by reference, when a PDE<T> must be a pointer in our current design anyway?

File: src/pde/pde_continuity1.hpp [hunk 40]
```diff
@@ -112,28 +112,33 @@ class PDE_continuity_1d : public PDE<P>
...
+      dimension<P>(-1.0,                   // domain min
+                   1.0,                    // domain max
+                   2,                      // levels
+                   2,                      // degree
+                   initial_condition_dim0, // initial condition
+                   "x");                   // name
 
   inline static std::vector<dimension<P>> const dimensions_ = {dim0_};
 
+  // clang-format off
```
Comment (grahamlopez) [inline src/pde/pde_continuity1.hpp]: remove clang-format comments

File: src/coefficients.cpp (hunk 39 above)
Comment (bmcdanie) [reply src/coefficients.cpp]: I asked for this. it's stored in a `unique_ptr`, but passing around pointers (even smart ones) is more cumbersome to me than deref'ing at callsites and passing the pde by reference. this is the pattern we use for other functions requiring `pde` be passed in from main.

File: src/pde/pde_continuity1.hpp (hunk 40 above)
Comment (bmcdanie) [reply src/pde/pde_continuity1.hpp]: done

File: src/coefficients.cpp (hunk 39 above)
Comment (grahamlopez) [reply src/coefficients.cpp]: (capitulating to off-line discussion. We've got bigger fish to fry)
